### ISSUE TITLE
Upgrading from 1.7.3 to 1.7.4 (a patch release) broke my service due to introduction of a "new" dependency

### ISSUE THREAD

**Issue by DavidRigglemanININ** [micrometer-metrics/micrometer#2895]
**Describe the bug**
When upgrading from 1.7.3 to 1.7.4, my spring boot based microservice started failing to boot up with the exception:
```
An attempt was made to call a method that does not exist. The attempt was made from the following location:

    io.micrometer.core.instrument.dropwizard.DropwizardMeterRegistry.newTimer(DropwizardMeterRegistry.java:105)

The following method did not exist:

    'com.codahale.metrics.Timer com.codahale.metrics.MetricRegistry.timer(java.lang.String, com.codahale.metrics.MetricRegistry$MetricSupplier)'
```
It appears this change was introduced by https://github.com/micrometer-metrics/micrometer/pull/2684

I was able to fix the issue by manually adding in my dependencies section of my pom.xml file (I'm using maven) an explicit dependency on 
```
<dependency>
      <groupId>com.codahale.metrics</groupId>
      <artifactId>metrics-core</artifactId>
      <version>3.0.2</version>
</dependency>
```
But given this is a very old and deprecated library, it seems very undesirable for this to be necessary and I'm not sure if it would cause any additional conflicts with the dropwizard version of the library also on my classpath. Am I misconfiguring something or doing something else wrong?

For additional context, I already had the following declared as a dependency but it wasn't good enough
```
<dependency>
    <groupId>io.dropwizard.metrics</groupId>
    <artifactId>metrics-core</artifactId>
    <version>4.2.4</version>
</dependency>
```

**Environment**
 - Micrometer version: >= 1.7.4
 - OS: macOS
 - Java version: 11

**To Reproduce**
How to reproduce the bug:
Upgraded from 1.7.3 to 1.7.4

**Expected behavior**
A patch release should not introduce breaking changes and also not be forced to add a dependency on a very ancient library (latest version released in 2014)

**Comment by jonatan-ivanov** [comment]
This is not a new dependency, `DropwizardMeterRegistry` has always had a dependency on dropwizard-core (see [here](https://github.com/micrometer-metrics/micrometer/blob/64dba5be59306ae8604340b45bcabe31d9147428/micrometer-core/src/main/java/io/micrometer/core/instrument/dropwizard/DropwizardMeterRegistry.java#L19-L20)).
You can find that `micrometer-core` has an [optional](https://github.com/micrometer-metrics/micrometer/blob/64dba5be59306ae8604340b45bcabe31d9147428/micrometer-core/build.gradle#L13) dependency on `io.dropwizard.metrics:metrics-core`.
Also, 1.7.x [depends on](https://github.com/micrometer-metrics/micrometer/blob/1.7.x/micrometer-core/gradle.lockfile#L54) `io.dropwizard.metrics:metrics-core:4.0.7` not on 3.x.

I guess this is a classpath issue in your app, please check your dependencies and upgrade micrometer to 1.7.6 or better: 1.8.0.
(I'm going to release today so 1.7.7 and 1.8.1 hopefully soon.)

I'm going to close this because of the above, please let us know if this is incorrect/the issue persists and we can reopen.

**Comment by shakuzen** [comment]
> When upgrading from 1.7.3 to 1.7.4, my spring boot based microservice started failing to boot up

If you're using Micrometer with Spring Boot, it is generally a good idea to let Spring Boot's dependency management determine the version of Micrometer dependencies that get used.

**Comment by DavidRigglemanININ** [comment]
I already had a dependency on io.dropwizard.metrics:metrics-core in my project (actually 4.2.4) and that wasn't good enough as I was still getting the error. I also had to resolve the micrometer dependency manually because I have multiple libraries (besides just Spring Boot) that make use of this library. So it's not possible to simply let Spring Boot determine the version as nice as that may be. So I guess where I'm still confused is why manually adding the 3.0.2 dependency fixed things.

**Comment by DavidRigglemanININ** [comment]
Also, I did test with the latest (1.7.6 and 1.8.0) and the issue still persisted. Looking more into potential class path issues, I am going to try to exclude the old codahale metrics library that is a transitive dependency of another library I'm bringing in.

**Comment by jonatan-ivanov** [comment]
The process of classloading is pretty serious black magic. The decision about which class will be loaded if there are multiple ones on your classpath depends on your java version, your arch, your OS, your classpath, if it is Thursday, the weather on Mars and fluctuations in the Cosmic Microwave Background radiation. :D Changing the classpath might also change what is loaded and what is not even if the change seems unrelated/weird.

1.7.3, 1.7.4, 1.7.7 and 1.8.1 all have the same lines in their [pom.xml](https://repo1.maven.org/maven2/io/micrometer/micrometer-core/1.8.1/micrometer-core-1.8.1.pom) regarding dropwizard:

```xml
<dependency>
  <groupId>io.dropwizard.metrics</groupId>
  <artifactId>metrics-core</artifactId>
  <version>4.0.7</version>
  <scope>compile</scope>
  <optional>true</optional>
</dependency>
```

This is how you can troubleshoot this:
1. Check your classpath for instances of `io.dropwizard.metrics` (`gradle dependencies`, `mvn dependency:tree`)
2. Ensure that you have one version from all of the dropwizard artifacts (force/exclude them in gradle/maven)
3. Make sure that the version you are using has the method that you see in the error: you can decompile your jar using your IDE or download the sources (decompile is less error prone)
I assume that all 4.x versions should have that but please double check

If this does not help and you still see the problem, could you please create a minimal sample application that can reproduce the issue? I can help and take a look.

**Comment by DavidRigglemanININ** [comment]
Thanks for those tips. I was able to finally get it working and you are right both that it was a classpath issue and that it's also some pretty serious black magic. I think that by adding the old library in my pom explicitly, it actually shifted the classpath ordering to where it was now below the more current version of the package. So basically, adding a dependency actually caused me not to use it. Go figure. But I guess this is what happens when you create a new artifact with the same package. Anyhow, I used dependency:tree to add the necessary exclusions, which oddly enough I had to run multiple times before all usages showed up as new ones kept popping up after removing other usages).
